### PR TITLE
Support for AI.DAGRUN and AI.DAGRUN_RO

### LINKED ISSUE #7: support for AI.DAGRUN

### LINKED ISSUE #8: support for AI.DAGRUN_RO

### PR DESCRIPTION
**Pull request by filipecosta90**
This PR adds support for DAGRUN and DAGRUN_RO commands ( including the newly added SCRIPTRUN within the DAG ).
- fixes #7 
- fixes #8 
- adds `scriptRunFlatArgs`, `modelRunFlatArgs`, `tensorGetFlatArgs` to reuse code across Dag and RedisAI files
- rename `getTensorSetCommandBytes` to `tensorSetFlatArgs` to have the same signature specs as other command arg builders

### CODE REVIEW

**Comment by filipecosta90** [comment]
@dengliming can you also share your thoughts? I was not able to add you as a reviewer but would appreciate your comments on it =)

File: src/main/java/com/redislabs/redisai/Dag.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
+package com.redislabs.redisai;
+
+import java.util.ArrayList;
+import java.util.List;
+import redis.clients.jedis.util.SafeEncoder;
+
+public class Dag implements DagRunCommands {
+  List<List<byte[]>> commands;
+  List<Boolean> tensorgetflag;
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisai/Dag.java:9]:
```
private final  List<List<byte[]>> commands = new ArrayList<>();
private final  List<Boolean> tensorgetflag = new ArrayList<>();
```

File: src/main/java/com/redislabs/redisai/Dag.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  @Override
+  public void runScript(String key, String function, String[] inputs, String[] outputs) {
+    List<byte[]> args = Script.scriptRunFlatArgs(key, function, inputs, outputs, true);
+    this.commands.add(args);
+    this.tensorgetflag.add(false);
+  }
+
+  List<byte[]> dagRunFlatArgs(String[] loadKeys, String[] persistKeys) {
+    List<byte[]> args = new ArrayList<>();
+    if (loadKeys != null && loadKeys.length > 0) {
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisai/Dag.java:59]: !loadKeys.isEmpty()

File: src/main/java/com/redislabs/redisai/Dag.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    }
+    if (persistKeys != null && persistKeys.length > 0) {
+      args.add(Keyword.PERSIST.getRaw());
+      args.add(SafeEncoder.encode(String.valueOf(persistKeys.length)));
+      for (String key : persistKeys) {
+        args.add(SafeEncoder.encode(key));
+      }
+    }
+    for (List<byte[]> command : this.commands) {
+      args.add(SafeEncoder.encode("|>"));
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisai/Dag.java:74]: Replace `SafeEncoder.encode("|>")`
With a constant Keyword

File: src/main/java/com/redislabs/redisai/DagRunCommands.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package com.redislabs.redisai;
+
+public interface DagRunCommands {
+  public void setTensor(String key, Tensor tensor);
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisai/DagRunCommands.java:4]: interface methods don't need `public`

File: src/main/java/com/redislabs/redisai/Dag.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    this.tensorgetflag = new ArrayList<>();
+  }
+
+  protected List<?> processDagReply(List<?> reply) {
+    List<Object> outputList = new ArrayList<>(reply.size());
+    for (int i = 0; i < reply.size(); i++) {
+      if (this.tensorgetflag.get(i) == true) {
+        outputList.add(Tensor.createTensorFromRespReply((List<?>) reply.get(i)));
+      } else {
+        outputList.add(reply.get(i));
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisai/Dag.java:23]: `         outputList.add(reply.get(i)); `
What is the return value in this case?

File: src/test/java/com/redislabs/redisai/DagTest.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+      conn.flushAll();
+    }
+  }
+  //
+  //    @After
+  //    public void tearDown(){
+  //        try (Jedis conn = pool.getResource()) {
+  //            conn.flushAll();
+  //        }
+  //    }
```
Comment (gkorland) [inline src/test/java/com/redislabs/redisai/DagTest.java:26]: no need for tearDown?

File: src/main/java/com/redislabs/redisai/Dag.java (hunk 2 above)
Comment (filipecosta90) [reply src/main/java/com/redislabs/redisai/Dag.java:59]: using ArrayUtils given that this is in an array: `ArrayUtils.isNotEmpty(loadKeys)`

File: src/test/java/com/redislabs/redisai/DagTest.java (hunk 6 above)
Comment (filipecosta90) [reply src/test/java/com/redislabs/redisai/DagTest.java:26]: missed uncommenting it. solved in latest commit

File: src/main/java/com/redislabs/redisai/Dag.java (hunk 5 above)
Comment (filipecosta90) [reply src/main/java/com/redislabs/redisai/Dag.java:23]: String "OK"

**Comment by filipecosta90** [comment]
@gkorland I believe we're ready for re-review

File: src/main/java/com/redislabs/redisai/Dag.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  private final List<List<byte[]>> commands = new ArrayList<>();
+  private final List<Boolean> tensorgetflag = new ArrayList<>();
+
+  /** Direct acyclic graph of operations to run within RedisAI */
+  public Dag() {}
+
+  protected List<?> processDagReply(List<?> reply) {
+    List<Object> outputList = new ArrayList<>(reply.size());
+    for (int i = 0; i < reply.size(); i++) {
+      if (this.tensorgetflag.get(i) == true) {
```
Comment (dengliming) [inline src/main/java/com/redislabs/redisai/Dag.java:18]: Can be simplified to ` if (this.tensorgetflag.get(i))`

File: src/main/java/com/redislabs/redisai/Dag.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        outputList.add(Tensor.createTensorFromRespReply((List<?>) reply.get(i)));
+      } else {
+        outputList.add(reply.get(i));
+      }
+    }
+    return outputList;
+  }
+
+  @Override
+  public void setTensor(String key, Tensor tensor) {
```
Comment (dengliming) [inline src/main/java/com/redislabs/redisai/Dag.java:28]: Can we change to chained calls here? like: `public Dag setTensor(String key, Tensor tensor) ` end with `return this;` The same goes for other methods. : )

File: src/main/java/com/redislabs/redisai/DagRunCommands.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package com.redislabs.redisai;
+
+interface DagRunCommands {
```
Comment (dengliming) [inline src/main/java/com/redislabs/redisai/DagRunCommands.java:3]: If we need a chain call. We can change interface like this:
```
package com.redislabs.redisai;

interface DagRunCommands<T> {
  T setTensor(String key, Tensor tensor);

  T getTensor(String key);

  T runModel(String key, String[] inputs, String[] outputs);

  T runScript(String key, String function, String[] inputs, String[] outputs);
}
```
`public class Dag implements DagRunCommands<Dag>` 

WDYT?  : )

File: src/main/java/com/redislabs/redisai/Dag.java (hunk 8 above)
Comment (filipecosta90) [reply src/main/java/com/redislabs/redisai/Dag.java:28]: LvGTM :) changing

File: src/main/java/com/redislabs/redisai/DagRunCommands.java (hunk 9 above)
Comment (filipecosta90) [reply src/main/java/com/redislabs/redisai/DagRunCommands.java:3]: agree :+1: . applying changes

File: src/main/java/com/redislabs/redisai/Dag.java (hunk 7 above)
Comment (filipecosta90) [reply src/main/java/com/redislabs/redisai/Dag.java:18]: :+1:

File: src/main/java/com/redislabs/redisai/Keyword.java [hunk 10]
```diff
@@ -16,12 +15,19 @@
...
   private final byte[] raw;
 
   Keyword() {
-    raw = SafeEncoder.encode(this.name());
+    if (this.name() == "PIPE") {
+      raw = SafeEncoder.encode("|>");
+    } else {
+      raw = SafeEncoder.encode(this.name());
+    }
   }
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisai/Keyword.java:31]:
```java
PIPE("|>");

private final byte[] raw; 

Keyword() {
  this(this.name());
}

Keyword(String word) {
       raw = SafeEncoder.encode("|>"); 
} 
```

Comment (filipecosta90) [reply src/main/java/com/redislabs/redisai/Keyword.java:31]: @gkorland that implies:
```

public enum Keyword implements ProtocolCommand {
  INPUTS("INPUTS"),
  OUTPUTS("OUTPUTS"),
  META("META"),
  VALUES("VALUES"),
  BLOB("BLOB"),
  SOURCE("SOURCE"),
  RESETSTAT("RESETSTAT"),
  TAG("TAG"),
  BATCHSIZE("BATCHSIZE"),
  MINBATCHSIZE("MINBATCHSIZE"),
  BACKENDSPATH("BACKENDSPATH"),
  LOADBACKEND("LOADBACKEND"),
  LOAD("LOAD"),
  PERSIST("PERSIST"),
  PIPE("|>");

  private final byte[] raw;

  Keyword(String s) {
    raw = SafeEncoder.encode(s);
  }
(...)
```

Comment (dengliming) [reply src/main/java/com/redislabs/redisai/Keyword.java]: I think his suggestion is that you should add two constructors. :)
The complete code should be:
```
public enum Keyword implements ProtocolCommand {
  INPUTS,
  OUTPUTS,
  META,
  VALUES,
  BLOB,
  SOURCE,
  RESETSTAT,
  TAG,
  BATCHSIZE,
  MINBATCHSIZE,
  BACKENDSPATH,
  LOADBACKEND,
  LOAD,
  PERSIST,
  PIPE("|>");

  private final byte[] raw;

 Keyword() {
   this(this.name());
 }

  Keyword(String s) {
    raw = SafeEncoder.encode(s);
  }
(...)
```

Comment (filipecosta90) [reply src/main/java/com/redislabs/redisai/Keyword.java]: Your totally right @gkorland and @dengliming . was completely missing it =) now fixed

**Pull request merged**